New Slate optimisation pass

sv2518 opened the pull request:
This PR introduces a new optimisation pass for the Slate compiler.

In the new optimisation pass Slate expressions are resorted so that the corresponding generated code contains a minimal amount of matrix-shaped temporaries. This is useful for high order FEM where the local assembly tensors are big. By using less matrix-shaped temporaries we reduce the amount of data transfer, which is a bottleneck for high-order FEM. Note that the tradeoff is the execution of more FLOPS.

The resorting works by pushing the multiplication of a matrix with a vector as far inside an expression as possible. For example if we have an expressions `(A+B)*f`, where A,B are `Tensors` of rank 2 and f is an `AssembledVector`, then the new compiler pass resorts the expression to `A*f+B*f`. In the compiler pass we also replace all inverses with solves with help of `A.inv*f=A.solve(f)`.

While I want this compiler pass so that I can replace the multiplication in `A*f+B*f` by Actions later on, we should already benefit from this optimisation without that replacement. That is so because: 
1) E.g. `(A+B)*f` theoretically builds 3 matrix-shaped temporaries A, B and A+B, while `A*f+B*f` only builds A and B. BUT in the generated C code the expression is inlined (see C code below) and I don't exactly know what the C compiler is doing under the hood with that. I am afraid it is smart enough to optimise this expression, so that we maybe only really benefit from this optimisation step WITH the replacement of the multiplication by an Action. I have looked at the assembly code for this example, but I have no experience with it and don't understand it well enough to understand if what I say is true. I have timed the difference of the optimised and non-optimised expression for a FE discretisation that results in local tensors of size 400x400, but the timings are almost the same.
2) We certainly benefit of replacing inverse operations with solve operations. The assembly of the optimised expression on a warm cache takes a fraction of the non-optimised expression in this case.

An example of the generated C code is 
Not optimised expression `(A+B)*f`:
```
static void slate_loopy(double *__restrict__ output, double const *__restrict__ T0, double const *__restrict__ T1)
{
  for (int32_t i = 0; i <= 399; ++i)
    for (int32_t i_0 = 0; i_0 <= 399; ++i_0)
      output[i] = output[i] + (T0[400 * i + i_0] + T0[400 * i + i_0]) * T1[i_0];
}
```
Optimised expression `A*f+B*f`
```
static void slate_loopy(double *__restrict__ output, double const *__restrict__ T0, double const *__restrict__ T1)
{
  double t0;

  for (int32_t i = 0; i <= 399; ++i)
  {
    t0 = 0.0;
    for (int32_t i_0 = 0; i_0 <= 399; ++i_0)
      t0 = t0 + T0[400 * i + i_0] * T1[i_0];
    output[i] = output[i] + t0 + t0;
  }
}
```

In addition to the multiplication pass, I also introduce a transpose optimisation pass, which rewrites `A.T.T->A` wherever possible.

dham commented on firedrake/slate/slac/optimise.py:
```diff
@@ -83,3 +104,254 @@ def _push_block_block(expr, self, indices):
+
+@_drop_double_transpose.register(Factorization)
+def _drop_double_transpose_factorization(expr, self):
+    """ Drop any factorisations. """
```
Explain why this is happening here.

dham commented on firedrake/slate/slac/optimise.py:
```diff
@@ -83,3 +104,254 @@ def _push_block_block(expr, self, indices):
+        else:
+            return expr
+    else:
+        assert "Actions in Slate are not yet supported."
```
`assert False,`

dham commented on firedrake/slate/slac/optimise.py:
```diff
@@ -83,3 +104,254 @@ def _push_block_block(expr, self, indices):
+        :returns: an optimised transposed expression
+    """
+    if expr.rank == 2:
+        transposed_state = ActionBag(Transpose(state.coeff), state.pick_op ^ 1)  # switch the multiplication order with pick_op
```
```
(state.pick_op + 1) % 2
```

dham commented on firedrake/slate/slac/optimise.py:
line length.

sv2518 commented:
Okay, I addressed all comments. The failing test was caused because I screwed up cherry-picking the changes from my other PR (the one that we merged last week.) So this is resolved now.
